packages/langium/src/parser/langium-parser.ts

```diff
@@ -4,365 +4,212 @@
  * terms of the MIT License, which is available in the project root.
  ******************************************************************************/
 
-/* eslint-disable @typescript-eslint/no-explicit-any */
-import { EmbeddedActionsParser, ILexingError, IOrAlt, IRecognitionException, IToken, Lexer, TokenType } from 'chevrotain';
-import { AbstractElement, Action, isAssignment, isCrossReference } from '../grammar/generated/ast';
-import { LangiumDocument } from '../documents/document';
-import { AstNode, CompositeCstNode, CstNode, LeafCstNode, Reference } from '../syntax-tree';
-import { isArrayOperator } from '../grammar/grammar-util';
-import { CompositeCstNodeImpl, CstNodeBuilder, LeafCstNodeImpl } from './cst-node-builder';
-import { GrammarAccess } from '../grammar/grammar-access';
-import { Linker } from '../references/linker';
+import { TokenType } from 'chevrotain';
+import { AbstractElement, Action, Alternatives, CrossReference, Grammar, Group, isAction, isAlternatives, isAssignment, isCrossReference, isGroup, isKeyword, isParserRule, isRuleCall, isTerminalRule, isUnorderedGroup, Keyword, ParserRule, RuleCall, UnorderedGroup } from '../grammar/generated/ast';
+import { Cardinality, getTypeName, isArrayOperator, isDataTypeRule } from '../grammar/grammar-util';
 import { LangiumServices } from '../services';
-import { getContainerOfType } from '../utils/ast-util';
-import { ValueConverter } from './value-converter';
-
-export type ParseResult<T = AstNode> = {
-    value: T,
-    parserErrors: IRecognitionException[],
-    lexerErrors: ILexingError[]
+import { getContainerOfType, streamAllContents } from '../utils/ast-util';
+import { stream } from '../utils/stream';
+import { DatatypeSymbol, LangiumBaseParser } from './langium-base-parser';
+
+type RuleContext = {
+    option: number,
+    consume: number,
+    subrule: number,
+    many: number,
+    or: number
 }
 
-export const DatatypeSymbol = Symbol('Datatype');
-
-type RuleResult = () => any;
+type Method = () => void;
 
-export class LangiumParser {
-    readonly grammarAccess: GrammarAccess;
+export class LangiumParser extends LangiumBaseParser {
```

Comment:
I think we should use composition instead of inheritance here. The two classes have different responsibilities..

Reply:
You're right. I changed it that way 👍